### PR TITLE
Add utility upon death and average utility per agent type

### PR DESCRIPTION
**Pull request by daryllimyt**
https://github.com/SOMAS2021/SOMAS2021/issues/257

4 new metrics for base agent:
- totalFoodTaken food.FoodType
- totalFoodSeen  food.FoodType
- totalHPGained  int
- totalHPLost    int

Currently, base utility is the product of these two: (please critique and add comments, this probably isn't final)
- HPGainFactor - calculated using either 
    - Ratio: gain/loss and 1 + gain/maxhp (when loss == 0)
    - Net HP: 1 + (gain - loss) / maxhp, limited to values >= 0 only
- foodIntakeFactor
    - Ratio: 0 when total seen == 0, intake / total seen otherwise

This base utility value is then passed into another function to make utility levels context sensitive i.e. higher levels less marginal utility gain etc. (but I'm not really sure if I'm doing this right, or if the previous steps make sense. Kind of just swinging blindly here). Currently I just have a quadratic function as the final utility function:
image

Frontend looks something like this:
image

### CODE REVIEW

**Review by jzzheng22 — COMMENTED** [review]
LGTM. Not looking at the logging code - I didn't write that so don't feel comfortable commenting on it.

The utility code is fine.

Haven't looked at frontend either (code or UI).

Agree the utility function should probably be improved, but fine for now. (CC @MattScottEEE who will have thoughts on this)

File: pkg/infra/baseAgent.go [hunk 1]
```diff
@@ -210,10 +233,12 @@ func (a *Base) TakeFood(amountOfFood food.FoodType) (food.FoodType, error) {
 	if amountOfFood < 0 {
 		return 0, &NegFoodError{}
 	}
+	a.totalFoodSeen += a.tower.currPlatFood
```
Comment (JaafarRammal) [inline pkg/infra/baseAgent.go:236]: This is called:
- Once per tick
- Only if agent is calling `TakeFood`

Which means that if the agent never tried to take food we are not logging that they are seeing this food

File: pkg/infra/tower.go [hunk 2]
```diff
@@ -100,6 +100,7 @@ func (t *Tower) endOfDay() {
 	t.dayInfo.CurrDay += 1
 	for _, agent := range t.Agents {
 		agent := agent.BaseAgent()
+		t.stateLog.LogUtility(t.dayInfo, agent.agentType, agent.Utility(), agent.IsAlive())
```
Comment (JaafarRammal) [inline pkg/infra/tower.go:103]: I think `IsAlive` is updated after hp decay I called on the line after

File: pkg/infra/baseAgent.go [hunk 3]
```diff
@@ -226,6 +251,77 @@ func (a *Base) TakeFood(amountOfFood food.FoodType) (food.FoodType, error) {
 	return foodTaken, nil
 }
 
+func (a *Base) Utility() float64 {
```
Comment (JaafarRammal) [inline pkg/infra/baseAgent.go:254]: I'd prefer lower case `utility` since this is only called in the same package by tower

**Review by JaafarRammal — COMMENTED** [review]
Frontend code LGTM
Logging code LGTM
I think on the backend agent death check is being called too early?

File: pkg/infra/tower.go (hunk 2 above)
Comment (daryllimyt) [reply pkg/infra/tower.go:103]: I see I see, will change

File: pkg/infra/baseAgent.go (hunk 3 above)
Comment (daryllimyt) [reply pkg/infra/baseAgent.go:254]: sure np

File: pkg/infra/baseAgent.go (hunk 1 above)
Comment (daryllimyt) [reply pkg/infra/baseAgent.go:236]: I've moved this into a function that gets called on each tick and occurs before each agent takes action

**Review by JaafarRammal — DISMISSED** [review]

**Review by JaafarRammal — APPROVED** [review]

**Pull request merged**